llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -18923,7 +18919,7 @@ SDValue DAGCombiner::visitFP_ROUND(SDNode *N) {
     // single-step fp_round we want to fold to.
     // In other words, double rounding isn't the same as rounding.
     // Also, this is a value preserving truncation iff both fp_round's are.
-    if (DAG.getTarget().Options.UnsafeFPMath || N0IsTrunc)
+    if (N->getFlags().hasAllowContract() || N0IsTrunc)
```

Comment:
I don't think contract fits this. Unfortunately instcombine does not do the same fold for reference. The rounding flag is also weird, it should really be its own fast math flag instead of this special operand. 

I think this is more of an approximate function